Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp

```diff
@@ -0,0 +1,680 @@
+    if (!(hints & WPE_INPUT_HINT_INHIBIT_OSK)) {
+        zwp_text_input_v1_show_input_panel(global->textInput);
+        zwp_text_input_v1_activate(global->textInput,
+            seat->seat(),
+            wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(self->priv->view)));
+    }
+}
+
+static void wpeIMContextWaylandV1FocusOut(WPEInputMethodContext *context)
+{
+    auto* self = WPE_IM_CONTEXT_WAYLAND_V1(context);
+    auto* display = wpe_im_context_get_display(context);
+    auto* seat = wpeDisplayWaylandGetSeat(WPE_DISPLAY_WAYLAND(display));
+    auto* global = textInputV1GetGlobalByContext(self);
+    if (global->current == nullptr)
+        return;
+
+    if (global->focused)
+        zwp_text_input_v1_deactivate(global->textInput, seat->seat());
+
+    global->current = nullptr;
+}
+
+static void wpeIMContextWaylandV1SetCursorArea(WPEInputMethodContext* context, int x, int y, int width, int height)
+{
+    auto* self = WPE_IM_CONTEXT_WAYLAND_V1(context);
+    auto* display = wpe_im_context_get_display(context);
+    auto* global = textInputV1GetGlobalByDisplay(WPE_DISPLAY_WAYLAND(display));
+    if (self->priv->cursor_rect.x == x && self->priv->cursor_rect.y == y && self->priv->cursor_rect.width == width && self->priv->cursor_rect.height == height)
+        return;
+
+    self->priv->cursor_rect.x = x;
+    self->priv->cursor_rect.y = y;
+    self->priv->cursor_rect.width = width;
+    self->priv->cursor_rect.height = height;
+
+    if (global->current == context) {
+        textInputV1SetCursorRectangle(self);
+        textInputV1CommitState(self);
+    }
+}
+
+static void wpeIMContextWaylandV1SetSurrounding(WPEInputMethodContext* context, const char* text, guint length, guint cursorIndex, guint selectionIndex)
+{
+    auto* self = WPE_IM_CONTEXT_WAYLAND_V1(context);
+    auto* display = wpe_im_context_get_display(context);
+    auto* global = textInputV1GetGlobalByDisplay(WPE_DISPLAY_WAYLAND(display));
+
+    g_clear_pointer(&self->priv->surrounding.text, g_free);
+    self->priv->surrounding.text = g_strndup(text, length);
+    self->priv->surrounding.cursorIndex = cursorIndex;
+    self->priv->surrounding.anchorIndex = selectionIndex;
+
+    if (global->current == context) {
+        textInputV1SetSurroundingText(self);
+        textInputV1CommitState(self);
+    }
+}
+
+static void wpeIMContextWaylandV1Reset(WPEInputMethodContext *context)
+{
+    auto* self = WPE_IM_CONTEXT_WAYLAND_V1(context);
+    auto* global = textInputV1GetGlobalByContext(self);
+    if (global->current == nullptr)
+        return;
+
+    if (global->current == context) {
+        zwp_text_input_v1_reset(global->textInput);
+        textInputV1UpdateState(self);
+    }
+}
+
+static void wpe_im_context_wayland_v1_class_init(WPEIMContextWaylandV1Class* klass)
+{
+    GObjectClass* objectClass = G_OBJECT_CLASS(klass);
+    objectClass->constructed = wpeIMContextWaylandV1Constructed;
+    objectClass->finalize = wpeIMContextWaylandV1Finalize;
```

Comment:
Use dispose, finalize is already set by WEBKIT_DEFINE_TYPE to call the private struct destructor

Reply:
fixed

Reply:
I was not aware of this but why it's not ok to use finalize if I call parent finalize anyways?

Reply:
WEBKIT_DEFINE_TYPE macro already adds a finalize implementation that calls the private struct destructor. That allows us to use C++ types and smart pointers in GObject private structs allocated by glib. If you add your own the one defined by the macro is not used and the destructor not called.

Reply:
Fixed